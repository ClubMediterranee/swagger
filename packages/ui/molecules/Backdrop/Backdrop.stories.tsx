import {Meta, StoryFn} from "@storybook/react";

import {useState} from "react";

import {Backdrop} from "./Backdrop";

export default {
  title: "UI-Library/Component/Molecules/Backdrop",
  component: Backdrop,
  parameters: {
    layout: "fullscreen"
  }
} as Meta<typeof Backdrop>;

const Template: StoryFn<typeof Backdrop> = (args) => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <div>
      <button onClick={() => setIsOpen(true)}>
        <div
          className="h-screen w-screen object-cover"
          style={{
            backgroundImage: "url('https://fakeimg.pl/1000x1000')",
            backgroundSize: "cover"
          }}
        />
      </button>
      <Backdrop {...args} isVisible={isOpen} onClose={() => setIsOpen(false)}>
        <button
          onClick={() => {
            setIsOpen(false);
          }}
        >
          Close
        </button>
      </Backdrop>
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {};

export const Sweep_Mode = Template.bind({});
Sweep_Mode.args = {
  sweep: true
};
