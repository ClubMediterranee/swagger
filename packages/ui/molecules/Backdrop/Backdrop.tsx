import {animated, config, useTransition} from "@react-spring/web";
import classnames from "classnames";
import type {FunctionComponent, MouseEvent, PropsWithChildren} from "react";

interface BackdropProps {
  /**
   * On close
   * @param event
   */
  onClose?: (event: MouseEvent<HTMLButtonElement>) => void;
  /**
   * Is visible
   */
  isVisible?: boolean;
  /**
   * Sweep Modeâ„¢
   * (makes it appear from the sides instead of from everywhere)
   */
  sweep?: boolean;
}

export const Backdrop: FunctionComponent<PropsWithChildren<BackdropProps>> = ({
                                                                                children,
                                                                                onClose,
                                                                                isVisible,
                                                                                sweep
                                                                              }) => {
  const transitions = useTransition(isVisible, {
    from: {
      WebkitBackdropFilter: "blur(0px)",
      backdropFilter: "blur(0px)",
      "--tw-bg-opacity": 0,
      x: sweep ? "-100%" : "0%"
    },
    enter: {
      WebkitBackdropFilter: "blur(10px)",
      backdropFilter: "blur(10px)",
      "--tw-bg-opacity": 0.8,
      x: "0%"
    },
    leave: {
      WebkitBackdropFilter: "blur(0px)",
      backdropFilter: "blur(0px)",
      "--tw-bg-opacity": 0,
      x: sweep ? "100%" : "0%"
    },
    config: sweep ? {tension: 200, friction: 40} : config.gentle
  });

  return transitions(
    (style, item) =>
      item && (
        <div
          role="presentation"
          className="z-5 pointer-events-none bg-white fixed inset-0 flex items-center justify-center"
        >
          <animated.button
            className={classnames("-z-1 absolute inset-0 bg-white", {
              "pointer-events-auto": isVisible
            })}
            style={style}
            onClick={onClose}
          />
          {children}
        </div>
      )
  );
};
