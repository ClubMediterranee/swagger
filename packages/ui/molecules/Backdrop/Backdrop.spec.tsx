import {render, screen} from "@testing-library/react";

import {Backdrop} from "./Backdrop";

describe("<Backdrop />", () => {
  it("renders a Backdrop component", () => {
    render(
      <Backdrop isVisible>
        <div>backdrop content</div>
      </Backdrop>
    );
    const backdrop = screen.getByRole("presentation");
    expect(backdrop).toBeInTheDocument();
    expect(backdrop).toMatchSnapshot("backdrop");
  });
});
