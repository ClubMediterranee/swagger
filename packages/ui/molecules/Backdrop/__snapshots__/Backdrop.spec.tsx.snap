// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Backdrop /> renders a Backdrop component: backdrop 1`] = `
<div
  class="z-1 pointer-events-none fixed inset-0 flex items-center justify-center"
  role="presentation"
>
  <button
    class="-z-1 absolute inset-0 bg-white pointer-events-auto"
    style="--tw-bg-opacity: 0.8; transform: none;"
  />
  <div>
    backdrop content
  </div>
</div>
`;
